src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -80,8 +78,20 @@ private static object InvokeExpression(Expression expression)
             return Expression.Lambda(expression).Compile().DynamicInvoke();
         }
 
+        private static void CheckArgumentExpressionIsValid(Expression expression)
+        {
+            var visitor = new ArgumentConstraintExpressionVisitor();
+            visitor.Visit(expression);
+            if (visitor.ConstraintExpressions.Any(e => e != expression))
+            {
+                throw new InvalidOperationException(ExceptionMessages.ArgumentConstraintCannotBeNestedInArgument);

```

Comment:
`Expression.ToString()` returns things like ``A`1.That.Matches(x => True)``, which isn't great, but is consistent with the CLR formatting used elsewhere... But when we do the same from VB.NET, we get something like this: ``() => Convert(A`1.That).Matches(a0 => Invoke(() => True))``, which is much uglier.

The cheap alternative sounds reasonable to me.